onnxruntime/core/providers/shared_library/provider_author.h

```diff
@@ -0,0 +1,245 @@
+  TensorShape(const std::initializer_list<int64_t>& dims) : std::vector<int64_t>{dims} {}
+
+  TensorShape(const int64_t* dimension_sizes, size_t dimension_count);
+  TensorShape(const std::vector<int64_t>& dims, size_t start, size_t end);
+
+  using std::vector<int64_t>::operator[];
+
+  size_t NumDimensions() const noexcept {
+    return size();
+  }
+
+  const std::vector<int64_t>& GetDims() const { return *this; }
+
+  int64_t Size() const;
+
+  /**
+     Return a new TensorShape of the dimensions from dimstart to dimend.
+  */
+  TensorShape Slice(size_t dimstart, size_t dimend) const;
+
+  /**
+     Return a new TensorShape of the dimensions from dimstart to end.
+  */
+  TensorShape Slice(size_t dimstart) const;
+
+  /**
+     output dimensions nicely formatted
+  */
+  std::string ToString() const;
+
+  /**
+     Calculate size between start and end.
+     Assumes start and end are between 0 and this->NumDimensions(), inclusive, and that
+     start < end.
+  */
+  int64_t SizeHelper(size_t start, size_t end) const;
+};
+
+constexpr const char* kMSDomain = "com.microsoft";
+constexpr const char* kMklDnnExecutionProvider = "MKLDNNExecutionProvider";
+
+template <typename T>
+using IAllocatorUniquePtr = std::unique_ptr<T, std::function<void(T*)>>;
+
+std::unique_ptr<Provider_IDeviceAllocator> CreateCPUAllocator(std::unique_ptr<Provider_OrtMemoryInfo> memory_info);
+Provider_AllocatorPtr CreateDummyArenaAllocator(std::unique_ptr<Provider_IDeviceAllocator> resource_allocator);
+Provider_AllocatorPtr CreateAllocator(Provider_DeviceAllocatorRegistrationInfo& info, int16_t device_id = 0);
+
+class CPUIDInfo {
+ public:
+  static const CPUIDInfo& GetCPUIDInfo();
+
+  bool HasAVX2() const;
+  bool HasAVX512f() const;
+};
+
+namespace logging {
+
+enum class Severity {
+  kVERBOSE = 0,
+  kINFO = 1,
+  kWARNING = 2,
+  kERROR = 3,
+  kFATAL = 4
+};
+
+enum class DataType {
+  SYSTEM = 0,  ///< System data.
+  USER = 1     ///< Contains potentially sensitive user data.
+};
+
+struct Category {
+  static const char* onnxruntime;  ///< General output
+  static const char* System;       ///< Log output regarding interactions with the host system
+  // TODO: What other high level categories are meaningful? Model? Optimizer? Execution?
+};
+
+constexpr const char* SEVERITY_PREFIX = "VIWEF";
+
+class Logger {
+ public:
+  bool OutputIsEnabled(Severity severity, DataType data_type) const noexcept;
+};
+
+class LoggingManager {
+ public:
+  static const Logger& DefaultLogger();
+};
+
+class Capture {
+ public:
+  Capture(const Logger& logger, logging::Severity severity, const char* category,
+          logging::DataType dataType, const CodeLocation& location);
+
+  std::ostream& Stream() noexcept;
+};
+}  // namespace logging
+
+enum class AutoPadType {
+  NOTSET = 0,
+  VALID = 1,
+  SAME_UPPER = 2,
+  SAME_LOWER = 3,
+};
+
+inline AutoPadType StringToAutoPadType(const std::string& str) {
+  if (str.empty()) {
```

Comment:
This seems very specific to what an EP implements. Does it belong in a shared common header or should the EP implement/cut-and-paste this? If we do this here it creates a scenario where virtually any 'string to enum' helper for any op could be put here.

Reply:
Yeah, I'll probably make this be implemented through the host in the future. I'll put in a todo.